### PR TITLE
crypto/skip_old_blocks_when_restarting_the_node

### PR DESCRIPTION
**Pull request by Binary-Mohanad**

### CODE REVIEW

File: lib/Net/Async/Blockchain/ETH.pm [hunk 1]
```diff
@@ -231,7 +231,13 @@ sub subscribe {
         $self->recursive_search()->then(
             async sub {
                 while (1) {
-                    my $block_number = await $self->newHeads(await $self->new_blocks_queue->shift);
+                    # Skip the old chekced blocks
+                    # This could happen when restarting the node
```
Comment (turnstep) [inline lib/Net/Async/Blockchain/ETH.pm:235]:
```suggestion
                    # Skip old checked blocks (e.g. when restarting the node)
```

File: lib/Net/Async/Blockchain/ETH.pm [hunk 2]
```diff
@@ -231,7 +231,13 @@ sub subscribe {
...
+                    my $next_block        = await $self->new_blocks_queue->shift;
+                    my $next_block_number = Math::BigInt->from_hex($next_block->{params}->{result}->{number});
+                    next if $self->base_block_number and $next_block_number->blt($self->base_block_number);
```
Comment (turnstep) [inline lib/Net/Async/Blockchain/ETH.pm:238]: So $next_block is guaranteed to hold something?

Comment (Binary-Mohanad) [reply lib/Net/Async/Blockchain/ETH.pm:238]: This an example of what `$next_block` is hold

```code
$VAR1 = {
          'params' => {
                        'subscription' => '0x706ac971713561b1f953948ebad405e6',
                        'result' => {
                                      'sha3Uncles' => '0x1dcc4de8dec75d7aab85b567b6ccd41ad312451b948a7413f0a142fd40d49347',
                                      'nonce' => '0x0000000000000000',
                                      'hash' => '0xee5fcbeb374d48232728d661fba7f08365f7f17e4e3b360a544cdc29561b3bd1',
                                      'mixHash' => '0x0000000000000000000000000000000000000000000000000000000000000000',
                                      'miner' => '0x0000000000000000000000000000000000000000',
                                      'stateRoot' => '0xdd2bd75c0cd48e07757d4e6e13e5017a626297c8f0b96b70510a6cbab84e22b0',
                                      'difficulty' => '0x2',
                                      'transactionsRoot' => '0x56e81f171bcc55a6ff8345e692c0f86e5b48e01b996cadc001622fb5e363b421',
                                      'receiptsRoot' => '0x56e81f171bcc55a6ff8345e692c0f86e5b48e01b996cadc001622fb5e363b421',
                                      'parentHash' => '0xbbad625d17c310174c78d7e4767a781d700103075de62d0667e744fecdbcf8bb',
                                      'gasLimit' => '0x7a1200',
                                      'logsBloom' => '0x00000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000',
                                      'timestamp' => '0x5fc6f0b9',
                                      'gasUsed' => '0x0',
                                      'number' => '0x1d3379',
                                      'extraData' => '0xd683010914846765746886676f312e3135856c696e757800000000000000000028ebb639c1e31f5ef56db129fad0029d6add2e3af02d2038f7fe5ed241189a2870f4f1203f5cc0e4c1a553b73af9d9a02956ef9eccf7a6e58b81e33e142d0e3100'
                                    }
                      },
          'jsonrpc' => '2.0',
          'method' => 'eth_subscription'
        };
```

But so far, we just care about the block number

Comment (turnstep) [reply lib/Net/Async/Blockchain/ETH.pm:237]: Sorry, I meant does it always exist? For example, can new_blocks_queue be empty at this point?

Comment (Binary-Mohanad) [reply lib/Net/Async/Blockchain/ETH.pm:237]: We are using [Future::Queue](https://metacpan.org/pod/Future::Queue) and as mentioned in the package description, if the queue is empty then the future that `shift` returns will be completed once an item becomes available.

**Pull request merged**